### PR TITLE
[CG-10888] fix: Wildcard resolution

### PR DESCRIPTION
**Pull request by tkfoss**

### CODE REVIEW

**Comment by codegen-team** [comment]
CodegenBot is starting to review the PR please wait...

**Comment by codegen-team** [comment]
Overall Assessment:
The changes significantly improve wildcard import handling and symbol resolution in the codebase. The implementation is thorough and well-tested.

Strengths:
1. Comprehensive test coverage across various wildcard import scenarios
2. Clean implementation of recursive symbol resolution
3. Proper handling of nested module structures
4. Good attention to edge cases like circular imports

Minor Suggestions:
1. Consider adding docstrings to the new PyFile methods to maintain documentation consistency
2. The test file has a typo in function name: `test_import_wildcard_preserves_import_resultion` should be `resolution`
3. Consider adding debug logging for complex resolution chains to aid in troubleshooting

File: src/codegen/sdk/python/file.py [hunk 1]
```diff
@@ -197,3 +197,28 @@ def valid_import_names(self) -> dict[str, PySymbol | PyImport | WildcardImport[P
                         ret[file.name] = file
             return ret
         return super().valid_import_names
+
```
Comment (codegen-team) [inline src/codegen/sdk/python/file.py:200]: Consider adding docstring to explain the purpose and behavior of this method. Suggested format:
```python
def get_node_from_wildcard_chain(self, symbol_name: str):
    """Recursively resolves a symbol through wildcard import chains.
    
    Args:
        symbol_name: Name of the symbol to resolve
        
    Returns:
        The resolved node or None if not found
    """
```

File: tests/unit/codegen/sdk/python/import_resolution/test_import_resolution.py [hunk 2]
```diff
@@ -343,7 +543,7 @@ def some_func():
         assert len(some_func.symbol_usages) > 0
 
 
-def test_import_wildcard_preserves_import_resolution(tmpdir: str) -> None:
+def test_import_wildcard_preserves_import_resultion(tmpdir: str) -> None:
```
Comment (codegen-team) [inline tests/unit/codegen/sdk/python/import_resolution/test_import_resolution.py]: The test name has a typo: `resultion` should be `resolution`. Consider fixing for consistency.

File: src/codegen/sdk/python/file.py (hunk 1 above)
Comment (codegen-team) [inline src/codegen/sdk/python/file.py:200]: Consider adding type hints to the return value for better code maintainability:
```python
def get_node_from_wildcard_chain(self, symbol_name: str) -> Optional[PySymbol]:
```

File: src/codegen/sdk/python/import_resolution.py [hunk 3]
```diff
@@ -232,6 +239,33 @@
...
+            for name, wildcard_import in self.names:
+                if name in self._requesting_names:
+                    yield from [frame.parent_frame for frame in wildcard_import.resolved_type_frames]
+
+    @noapidoc
+    def set_requesting_names(self, requester: PyImport):
+        if requester.is_wildcard_import():
+            self._requesting_names.update(requester._requesting_names)
+        else:
+            self._requesting_names.add(requester.name)
```
Comment (codegen-team) [inline src/codegen/sdk/python/import_resolution.py]: Consider adding debug logging for complex resolution chains. This could help with troubleshooting issues in production:
```python
def set_requesting_names(self, requester: PyImport):
    logger.debug(f"Setting requesting names from {requester.name}")
    if requester.is_wildcard_import():
        self._requesting_names.update(requester._requesting_names)
    else:
        self._requesting_names.add(requester.name)
```

**Review by bagel897 — CHANGES_REQUESTED** [review]
Can we keep the implementation consistent between typescript and python? https://github.com/codegen-sh/codegen-sdk/blob/09578e213d20921bf105e3d87c06d477e8af83b8/src/codegen/sdk/core/import_resolution.py#L333 
Here's how typescript does it right now. 

My understanding of the current situation
```python
# file a
TEST_CONST = 1
TEST_CONST2 = 1
# file b
from filea import *
t = TEST_CONST # Works in python and typescript
# file c
from fileb import TEST_CONST # Doesn't work in python before this PR, should work in typescript afaik
# file d
from fileb import TEST_CONST2 # Would add a duplicate edge pointing to TEST_CONST in python, works in typescript afaik
```
Here's a test case that demonstrates the file d edge case.
```python
def test_import_resolution_init_wildcard(tmpdir: str) -> None:
    """Tests that named import from a file with wildcard resolves properly"""
    # language=python
    content1 = """TEST_CONST=2
    foo=9
    """
    content2 = """from testdir.test1 import *
    bar=foo
    test=TEST_CONST"""
    content3 = """from testdir import TEST_CONST
    test3=TEST_CONST"""
    content4 = """from testdir import  foo
    test4=foo"""
    with get_codebase_session(tmpdir=tmpdir, files={"testdir/test1.py": content1, "testdir/__init__.py": content2, "test3.py": content3, "test4.py": content4}) as codebase:
        file1: SourceFile = codebase.get_file("testdir/test1.py")
        file2: SourceFile = codebase.get_file("testdir/__init__.py")
        file3: SourceFile = codebase.get_file("test3.py")

        symb = file1.get_symbol("TEST_CONST")
        test = file2.get_symbol("test")
        test3 = file3.get_symbol("test3")
        test3_import = file3.get_import("TEST_CONST")

        assert len(symb.usages) == 3
        assert symb.symbol_usages == [test, test3, test3_import]

```

**Review by bagel897 — APPROVED** [review]
